freegs/reconstruction.py

```diff
@@ -0,0 +1,803 @@
+
+    def plot_filaments(self):
+
+        # Plots the eigenbasis for the machine
+        plt.figure()
+        for filament in self.tokamak.vessel:
+            for subfil in filament.filaments:
+                if subfil.current >= 0:
+                    plt.scatter(subfil.R, subfil.Z, color='red',
+                                s=abs(subfil.current))
+                else:
+                    plt.scatter(subfil.R, subfil.Z, color='black',
+                                s=abs(subfil.current))
+
+        plt.show()
+
+    def plot_equilibria(self, fig, axs):
+        # Plot state of plasma equilibrium
+        if self.pause < 0:
+            fig = plt.figure()
+            axis = fig.add_subplot(111)
+        else:
+            axs[0].clear()
+            axs[1].clear()
+
+        plotting.plotEquilibrium(self, axis=axs[0], show=False)
+        if self.Jtor is not None:
+            axs[1].imshow(np.rot90(self.Jtor), extent=(0.1, 2, -1, 1),
+                          cmap=plt.get_cmap('jet'))
+
+        if self.pause < 0:
+            # Wait for user to close the window
+            plt.show()
+        else:
+            # Update the canvas and pause
+            # Note, a short pause is needed to force drawing update
+            axs[0].figure.canvas.draw()
+            axs[1].figure.canvas.draw()
+            plt.pause(self.pause)
+
+    def print_reconstruction(self):
+        """
+        Print data from reconstruction
+        """
+
+        print('Completed in ', self.it, 'iterations, with Chi Squared =', self.chi)
+
+        if self.use_VerticalControl:
+            self.dz = self.get_index(self.c, 'plasma')[-1]
+
+        print('Vertical Control Parameter: dz = ', self.dz)
+
+        print(' ')
+        print('Coefficients (alpha_i, dz, coil currents, vessel basis coefficients)')
+        print(self.c)
+
+        # Printing final constructed values
+        print(' ')
+        print('Reconstruction Sensor Diagnostics')
+        self.tokamak.printMeasurements(equilibrium=self)
+        self.tokamak.printCurrents()
+
+        for sensor, val1, val2 in zip(self.tokamak.sensors, self.M, self.H):
+            print(sensor.name, val1, val2, sensor.measurement)
+
+        for (name, coil), val1, val2 in zip(self.tokamak.coils, self.M[-self.tokamak.n_coils:], self.H[-self.tokamak.n_coils:]):
+            print(name, val1, val2, coil.current)
+
+    # Methods for solving
+    def solve_from_tokamak(self):
+        self.take_Measurements_from_tokamak()
+        self.reconstruct()
+        self.print_reconstruction()
+        #self.plot(show=self.show)
+
+        if self.test:
+            self.H=None
+            self.M=None
+            self.Jtor=None
+            self.dz=None
+            self.chi=None
+            self.sigma=None
+            self.psi_norm=None
+            self.c=None
+            self.it=0
+            self.is_limited = False
```

Comment:
what is this? and why?

Reply:
this is just resetting all of the variables for testing, as running one reconstruction straight after the other within the same class requires a clean slate (you dont want to be starting one equilibria with the plasma_psi from another one